x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SingleValueQuery.java

```diff
@@ -344,6 +360,92 @@ protected AbstractBuilder rewrite(QueryBuilder next) {
         }
     }
 
+    /**
+     * Builds a query matching either ignored values OR the union of {@code next} query
+     * and {@link SingleValueMatchQuery}. Three total queries. This is used to generate
+     * candidate matches for queries like {@code NOT(a == "b")} where some values of {@code a}
+     * are not indexed. In fact, let's use that as an example.
+     * <p>
+     *     In that case you use a query for {@code a != "b"} as the "next" query. Then
+     *     this query will find all documents where {@code a} is single valued and
+     *     {@code == "b"} AND all documents that have ignored some values of {@code a}.
+     *     This produces <strong>candidate</strong> matches for {@code NOT(a == "b")}.
+     *     It'll find documents like:
+     * </p>
+     * <ul>
+     *     <li>"a"</li>
+     *     <li>ignored_value</li>
+     *     <li>["a", ignored_value]</li>
+     *     <li>[ignored_value1, ignored_value2]</li>
+     *     <li>["b", ignored_field]</li>
+     * </ul>
+     * <p>
+     *     The first and second of those <strong>should</strong> match {@code NOT(a == "b")}.
+     *     The last three should be rejected. So! When using this query you <strong>must</strong>
+     *     push this query to the {@link LuceneSourceOperator} <strong>and</strong>
+     *     retain it in the {@link FilterOperator}.
+     * </p>
+     * <p>
+     *     This will not find:
+     * </p>
+     * <ul>
+     *     <li>"b"</li>
+     * </ul>
+     * <p>
+     *     And that's also great! These can't match {@code NOT(a == "b")}
+     * </p>
+     */
+    public static class NegatedSyntheticSourceDelegateBuilder extends AbstractBuilder {
+        NegatedSyntheticSourceDelegateBuilder(QueryBuilder next, String field, Source source) {
+            super(next, field, source);
+        }
+
+        @Override
+        public String getWriteableName() {
+            throw new UnsupportedOperationException("Not serialized");
+        }
+
+        @Override
+        protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject("negated_" + ENTRY.name);
+            builder.field("field", field() + ":synthetic_source_delegate");
+            builder.field("next", next(), params);
+            builder.field("source", source().toString());
+            builder.endObject();
+        }
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            throw new UnsupportedOperationException("Not serialized");
+        }
+
+        @Override
+        protected final org.apache.lucene.search.Query doToQuery(SearchExecutionContext context) throws IOException {
+            MappedFieldType ft = context.getFieldType(field());
+            if (ft == null) {
+                return new MatchNoDocsQuery("missing field [" + field() + "]");
+            }
+            ft = ((TextFieldMapper.TextFieldType) ft).syntheticSourceDelegate();
```

Comment:
I think SemanticTextFieldType and MatchOnlyTextFieldType are text fields, but their base classes are not TextFieldMapper.TextFieldType.

Reply:
:+1:. Will look.

Reply:
Update on this one: we don't push either of those fields. I've expanded the tests to hit that as well. We can pick them up in follow-up changes.

Reply:
👍